### PR TITLE
Implement BlockManager.iset

### PR DESCRIPTION
**Pull request by jbrockmendel**
Preliminary to the PR that fixes `setitem_with_indexer` (#22036, #15686)

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -1087,7 +1087,7 @@ def value_getitem(placement):
             unfit_mgr_locs = np.concatenate(unfit_mgr_locs)
             unfit_count = len(unfit_mgr_locs)
 
-            new_blocks = []
+            new_blocks: List[Block] = []
```
Comment (simonjayhawkins) [inline pandas/core/internals/managers.py]: rather than add the variable annotation, does new_blocks need to be initialized outside the if else? make_block has the correct return type and the extend wouldn't be needed. (but I guess this mypy fixup is orthogonal to this PR)

File: pandas/core/internals/managers.py [hunk 2]
```diff
@@ -995,7 +995,21 @@ def delete(self, item):
         self._shape = None
         self._rebuild_blknos_and_blklocs()
 
-    def set(self, item, value):
+    def set(self, item: Label, value):
+        """
+        Set new item in-place. Does not consolidate. Adds new Block if not
+        contained in the current set of items
```
Comment (simonjayhawkins) [inline pandas/core/internals/managers.py]: minor nit. can you aim for one-liners for PEP 257compliance

**Review by jreback — APPROVED** [review]
lgtm. @simonjayhawkins comments & merge on green.

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/internals/managers.py]: I'm not a fan of type annotations when not required. but not a blocker.
